test/upchunk.spec.ts

```diff
@@ -0,0 +1,322 @@
+
+    upload.on('error', (err) => {
+      try {
+        expect(err.detail.chunk).to.equal(0);
+        expect(err.detail.attempts).to.equal(5);
+        done();
+      } catch (err) {
+        done(err);
+      }
+    });
+
+    upload.on('success', () => {
+      done(new Error(`Expected upload to fail due to failed attempts`));
+    });
+  });
+
+  it('chunkSuccess event is fired after each successful upload', (done) => {
+    let calls = 0;
+    xhrMock.put(endpoint, { status: 200 });
+
+    const upload = createUploadFixture();
+
+    upload.on('chunkSuccess', () => {
+      calls++;
+    });
+
+    upload.on('success', () => {
+      expect(calls).to.equal(2);
+      done();
+    });
+  });
+
+  /** @TODO Figure out how to best handle progress testing. See: https://www.npmjs.com/package/xhr-mock#upload-progress (CJP) */
+  // const isNumberArraySorted = (a: number[]): boolean => {
+  //   for (let i = 0; i < a.length - 1; i += 1) {
+  //     if (a[i] > a[i + 1]) {
+  //       return false;
+  //     }
+  //   }
+  //   return true;
+  // };
+
+  // it('progress event fires the correct upload percentage', (done) => {
+  //   const fileBytes = 1048576;
+  //   const upload = createUploadFixture(
+  //     {},
+  //     new File([new ArrayBuffer(fileBytes)], 'test.mp4')
+  //   );
+
+  //   const scopes = [
+  //     nock('https://example.com')
+  //       .matchHeader('content-range', `bytes 0-${fileBytes / 4 - 1}/${fileBytes}`)
+  //       .put('/upload/endpoint')
+  //       .reply(200),
+  //     nock('https://example.com')
+  //       .matchHeader(
+  //         'content-range',
+  //         `bytes ${fileBytes / 4}-${fileBytes / 2 - 1}/${fileBytes}`
+  //       )
+  //       .put('/upload/endpoint')
+  //       .reply(200),
+  //     nock('https://example.com')
+  //       .matchHeader(
+  //         'content-range',
+  //         `bytes ${fileBytes / 2}-${3 * fileBytes / 4 - 1}/${fileBytes}`
+  //       )
+  //       .put('/upload/endpoint')
+  //       .reply(200),
+  //     nock('https://example.com')
+  //       .matchHeader(
+  //         'content-range',
+  //         `bytes ${3 * fileBytes / 4}-${fileBytes - 1}/${fileBytes}`
+  //       )
+  //       .put('/upload/endpoint')
+  //       .reply(200),
+  //   ];
+
+  //   const progressCallback = jest.fn((percentage) => percentage);
+
+  //   upload.on('error', (err) => {
+  //     done(err);
+  //   });
+
+  //   upload.on('progress', (progress) => {
+  //     progressCallback(progress.detail);
+  //   });
+
+  //   upload.on('success', () => {
+  //     scopes.forEach((scope) => {
+  //       if (!scope.isDone()) {
+  //         done('All scopes not completed');
+  //       }
+  //     });
+  //     expect(progressCallback).toHaveBeenCalledTimes(4);
+  //     const progressPercentageArray = progressCallback.xhrMock.calls.map(([percentage]) => percentage);
+  //     expect(isNumberArraySorted(progressPercentageArray)).toBeTruthy();
+  //     done();
+  //   });
+  // }, 10000);
```

Comment:
I'm not feeling good about losing this test.

tbh I think we should figure out how to keep it before this gets merged

Reply:
@dylanjha cool I think that's reasonable. I'll do some additional investigation. Hopefully it doesn't entail another replacement for nock.